config/hcl2shim/values_test.go

```diff
@@ -232,6 +232,26 @@ func TestConfigValueFromHCL2Block(t *testing.T) {
 			&configschema.Block{},
 			nil,
 		},
+		// nulls should be filtered out of the config, since they couldn't exist
+		// in hcl.
+		{
+			cty.ObjectVal(map[string]cty.Value{
+				"list": cty.ListVal([]cty.Value{
+					cty.StringVal("ok"),
+					cty.NullVal(cty.String)}),
```

Comment:
What if the null is in the middle of a list? :thinking: 

I suppose just dropping nulls in the middle (as if they didn't exist at all) could be reasonable behavior, but it feels a little surprising.

It's a shame `ConfigValueFromHCL2` doesn't work in terms of a schema where we could prefer to instead insert a reasonable zero value that the SDK _would_ accept... but in the absense of that, I guess this is fine since as you say there was no such thing as a null before anyway and so this is only a concern for newly-written configurations.

Reply:
The referenced issue was because of a variable that defaulted to `null`. But yes, this should only apply to HCL2 config where you're inserting a `null` in a list somehow.

Reply:
I guess one worry here is that because this "drop nulls" is a legacy SDK behavior, switching to a newer SDK without these shims wouldn't do it, and so it would be implicitly a breaking change for anyone who was relying on this implicit removal.

Having said that then, I wonder if instead we should prefer to solve this by returning an error message to the user saying that null values are not allowed (I imagine we'd need to do that as a post-shim validation step since the shims themselves don't return errors) and then users won't come to depend on this behavior and be upset when it doesn't behave the same in other future SDKs.

Reply:
Yeah, I only followed the existing code here, which did the same for maps since there was no error to return. I'm going to leave this since it prevents panics for other callers, but I'm making a PR to catch these separately during validation.

Reply:
Ignoring nulls in _maps_ seems okay because it's consistent with our usual rule that `null` is equivalent to unset, but "unset" isn't a reasonable concept for list elements (leaving something out is the same as re-setting everything after it, because the indices must of course be contiguous), so I think an error is the better behavior here to avoid unfortunate surprises.